ykrt/src/trace/swt/mod.rs

```diff
@@ -0,0 +1,133 @@
+//! Software tracer.
+
+use crate::frame::BitcodeSection;
+
+use super::{errors::InvalidTraceError, AOTTraceIterator, TraceAction, TraceRecorder};
+use std::sync::Once;
+use std::{cell::RefCell, collections::HashMap, error::Error, ffi::CString, sync::Arc};
+
+static FUNC_NAMES_INIT: Once = Once::new();
+
+#[derive(Debug, Eq, PartialEq)]
+struct TracingBlock {
+    function_index: usize,
+    block_index: usize,
+}
+
+thread_local! {
+    static BASIC_BLOCKS: RefCell<Vec<TracingBlock>> = RefCell::new(vec![]);
+    static FUNC_NAMES: RefCell<HashMap<usize, CString>> = RefCell::new(HashMap::new());
+}
+
+extern "C" {
+    fn get_function_names(
+        section: *const BitcodeSection,
+        result: *mut *mut IRFunctionNameIndex,
+        len: *mut usize,
+    );
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct IRFunctionNameIndex {
+    pub index: usize,
+    pub name: *const libc::c_char,
+}
+/// Inserts LLVM IR basicblock metadata into a thread-local BASIC_BLOCKS vector.
+///
+/// # Arguments
+/// * `function_index` - The index of the function to which the basic block belongs.
+/// * `block_index` - The index of the basic block within the function.
+pub fn trace_basicblock(function_index: usize, block_index: usize) {
+    BASIC_BLOCKS.with(|v| {
+        v.borrow_mut().push(TracingBlock {
+            function_index,
+            block_index,
+        });
+    })
+}
+
+pub(crate) struct SWTracer {}
+
+impl super::Tracer for SWTracer {
+    fn start_recorder(self: Arc<Self>) -> Result<Box<dyn TraceRecorder>, Box<dyn Error>> {
+        BASIC_BLOCKS.with(|bbs| {
+            bbs.borrow_mut().clear();
+        });
+        Ok(Box::new(SWTTraceRecorder {}))
+    }
+}
+
+impl SWTracer {
+    pub fn new() -> Result<Self, Box<dyn Error>> {
+        Ok(SWTracer {})
+    }
+}
+
+struct SWTTraceRecorder {}
+
+impl TraceRecorder for SWTTraceRecorder {
+    fn stop(self: Box<Self>) -> Result<Box<dyn AOTTraceIterator>, InvalidTraceError> {
```

Comment:
We don't have to do so in this PR but the idea here is that `stop` should be *very* fast, and all the nasty slow stuff (iterating over things) is done in a custom `AOTTraceIterator`. I have a sort-of half-done version of hwt which does this, but it doesn't work yet (for boring reasons).

Reply:
updated 👉 9a18474546bdd172c52c86bd11858a87899eb59e

Reply:
Ah, I think 9a18474 might make things worse, in the sense that it's pushing much more pressure onto the normal interpreter? I guess this slows things down; or, at least, it will later.

I'm inclined to revert 9a18474 for now, and then do something a little different in a future PR. By then I hope I've raised my hwt PR which might make clearer what I was thinking of here (or it might not!).

Reply:
reverted 👉 68d876b0085a965a4e26a63323d718986ba97f89